core/src/main/java/org/apache/brooklyn/util/core/task/BasicExecutionContext.java

```diff
@@ -203,11 +211,46 @@ public Void apply(Task<?> it) {
             throw new IllegalArgumentException("Unhandled task type: task="+task+"; type="+(task!=null ? task.getClass() : "null"));
         }
     }
-    
+
     private void registerPerThreadExecutionContext() { perThreadExecutionContext.set(this); }
 
     private void clearPerThreadExecutionContext() { perThreadExecutionContext.remove(); }
 
+    private void checkUserSuppliedContext(Object task, Collection<Object> taskTags) {
+        Entity taskContext = BrooklynTaskTags.getWrappedEntityOfType(taskTags, BrooklynTaskTags.CONTEXT_ENTITY);
+        Entity defaultContext = BrooklynTaskTags.getWrappedEntityOfType(tags, BrooklynTaskTags.CONTEXT_ENTITY);
+        if (taskContext != null) {
+            if (log.isWarnEnabled()) {
+                String msg = "Deprecated since 0.10.0. Task " + task + " is submitted for execution but has context " +
+                        "entity (" + taskContext + ") tag set by the caller. ";
+                if (taskContext != defaultContext) {
+                    msg += "The context entity of the execution context (" + this + ") the task is submitted on is " +
+                            defaultContext + " which is different. This will cause any of them to be used at random at " +
+                            "runtime. ";
+                    if (task instanceof BasicTask) {
+                        msg += "Fixing the context entity to the latter. ";
+                    }
+                }
+                msg += "Setting the context entity by the caller is not allowed. See the documentation on " +
+                        "BrooklynTaskTags.tagForContextEntity(Entity) method for more details. Future Apache Brooklyn " +
+                        "releases will throw an exception instead of logging a warning.";
+
+                /**
+                 * @deprecated since 0.10.0
+                 */
+                // Should we rate limit?
+                log.warn(msg);
```

Comment:
Let's monitor (including grep'ing the output logs from the mvn build), and rate limit if we're seeing this a lot.